privacyidea/api/lib/prepolicy.py

```diff
@@ -131,9 +131,52 @@ def policy_wrapper(*args, **kwds):
         return policy_wrapper
 
 
+def set_random_pin(request=None, action=None):
+    """
+    This policy function is to be used as a decorator in the API setrandompin function
+    If the policy is set accordingly it adds a random PIN to the
+    request.all_data like.
+
+    It uses the policy ACTION.OTPPINSETRANDOM in SCOPE.ADMIN or SCOPE.USER to set a random OTP PIN
+    """
+    params = request.all_data
+    policy_object = g.policy_object
+    # Fixme: we need to allow passing the user
+    #user_object = get_user_from_param(params)
+    # get the length of the random PIN from the policies
+    role = g.logged_in_user.get("role")
+    username = g.logged_in_user.get("username")
+    if role == ROLE.ADMIN:
+        scope = SCOPE.ADMIN
+        admin_realm = g.logged_in_user.get("realm")
+        realm = params.get("realm", "")
+    else:
+        scope = SCOPE.USER
+        realm = g.logged_in_user.get("realm")
+        admin_realm = None
+    pin_pols = policy_object.get_action_values(action=ACTION.OTPPINSETRANDOM,
+                                               scope=scope,
+                                               adminrealm=admin_realm,
+                                               user=username,
+                                               realm=realm,
+                                               #user_object=user_object,
+                                               client=g.client_ip,
+                                               unique=True,
+                                               audit_data=g.audit_object.audit_data)
+
+    if len(pin_pols) == 0:
+        # We do this to avoid that an admin sets a random PIN manually!
+        raise TokenAdminError("You need to specify an admin policy 'otp_pin_set_random'.")
```

Comment:
For users, we need a user policy, not an admin policy. Should we just remove the word "admin" from this error message?